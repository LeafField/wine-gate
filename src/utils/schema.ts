import { z } from "zod";

export const searchSchema = z.object({
  search: z
    .string()
    .min(4, { message: "4文字以上入力してください" })
    .max(60, { message: "60文字以下で入力してください" }),
  category: z.string(),
  sort: z.enum(["new", "inexpensive", "popular"]),
});
