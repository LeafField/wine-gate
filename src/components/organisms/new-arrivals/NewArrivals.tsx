import React, { FC } from "react";
import { SmallArticleProps } from "../../../types/article_types";
import SmallArticle from "../../molecules/small-article/SmallArticle";
import Heading from "../../atoms/heading/Heading";

type Props = {
  articles: SmallArticleProps[];
};

const NewArrivals: FC<Props> = ({ articles }) => {
  return (
    <section className="desktop:w-[64.9375rem] laptop:w-[43.0625rem] tablet:w-[20.75rem] flex w-full flex-col gap-[3.75rem]">
      <div className="col-span-3">
        <Heading title="新着ワイン" />
      </div>
      <div className="desktop:grid-cols-3 laptop:grid-cols-2 grid grid-cols-1 justify-items-stretch gap-5">
        {articles.map((article) => (
          <SmallArticle key={article.id} {...article} />
        ))}
      </div>
    </section>
  );
};

export default NewArrivals;
