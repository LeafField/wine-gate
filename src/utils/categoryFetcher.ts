import { supabase } from "./supabase";
import { cache } from "react";
import { ArticleProps } from "../types/article_types";

export const categoryFetcher = cache(
  async (slug: string[]): Promise<ArticleProps[] | null> => {
    if (slug.length !== 3) return null;
    const category = slug[0];
    const sort = slug[1];
    const page = Number(slug[2]);
    const start = (page - 1) * 10;
    const end = page * 10 - 1;
    if (sort === "new") {
      const { data, error } = await supabase
        .from("wines")
        .select(
          "title,sober_or_sweet,tart,fruity,author_name,tags,id,price,image_src,categories(*)",
        )
        .eq("categories.category", category)
        .order("created_at", { ascending: false })
        .range(start, end);
      if (error) throw new Error("カテゴリの取得に失敗しました");
      return data;
    } else if (sort === "inexpensive") {
      const { data, error } = await supabase
        .from("wines")
        .select(
          "title,sober_or_sweet,tart,fruity,author_name,tags,id,price,image_src,categories(*)",
        )
        .eq("categories.category", category)
        .order("price", { ascending: true })
        .range(start, end);
      if (error) throw new Error("カテゴリの取得に失敗しました");
      return data;
    } else if (sort === "popular") {
      const { data, error } = await supabase
        .rpc("get_wine_category_favorite", {
          category_value: category,
        })
        .limit(10);
      if (error) throw new Error("カテゴリの取得に失敗しました");
      const articles: ArticleProps[] = data.map((article) => ({
        categories: {
          category: article.category_name,
          id: article.category_id,
          chara: article.category_chara,
          sub: article.category_sub,
        },
        author_name: article.author_name,
        fruity: article.fruity,
        id: article.id,
        image_src: article.image_src,
        price: article.price,
        sober_or_sweet: article.sober_or_sweet,
        tags: article.tags,
        tart: article.tart,
        title: article.title,
      }));

      return articles;
    }
    return null;
  },
);

export const categoryCountFetcher = cache(async (category: string) => {
  const { count } = await supabase
    .from("wines")
    .select("id,categories(category)", { count: "exact", head: true })
    .eq("categories.category", category);
  if (count) {
    const totalPage = Math.ceil(count / 10);
    return totalPage;
  } else {
    return 1;
  }
});
