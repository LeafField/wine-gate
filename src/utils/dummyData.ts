import dummyImage from "../images/dummy_wine.png";
import {
  SmallArticleProps,
  CategoryProps,
  WineProps,
  ArticleProps,
} from "../types/article_types";

export const smallArticleDummyData: SmallArticleProps[] = [
  {
    id: "1",
    title: "コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ",
    chara: "渋い",
    tags: "樽熟成",
    image_src: dummyImage.src,
  },
  {
    id: "2",
    title: "コノスル シャルドネ ビシクレタ・レゼルバ",
    chara: "辛口白",
    tags: "フルーティ",
    image_src: dummyImage.src,
  },
  {
    id: "3",
    title: "コノスル カルメネール ビシクレタ・レゼルバ",
    chara: "酸っぱい",
    tags: "樽熟成,濃厚",
    image_src: dummyImage.src,
  },
];

export const categoryDummyData: CategoryProps[] = [
  {
    id: 1,
    category: "sober",
    chara: "渋い",
    sub: "red",
  },
  {
    id: 2,
    category: "tart",
    chara: "酸っぱい",
    sub: "red",
  },
  {
    id: 3,
    category: "fruity",
    chara: "果実味豊か",
    sub: "red",
  },
  {
    id: 4,
    category: "white",
    chara: "辛口白",
    sub: "white",
  },
  {
    id: 5,
    category: "sweet",
    chara: "甘口白",
    sub: "white",
  },
  {
    id: 6,
    category: "rose",
    chara: "ロゼ",
    sub: "interim",
  },
  {
    id: 7,
    category: "orange",
    chara: "オレンジ",
    sub: "interim",
  },
  {
    id: 8,
    category: "sparkling",
    chara: "スパークリング",
    sub: "other",
  },
  {
    id: 9,
    category: "fortified",
    chara: "酒精強化",
    sub: "other",
  },
];

export const selectDummyData = categoryDummyData.map((category) => {
  return { value: category.category, label: category.chara };
});

export const wineDummyData: WineProps = {
  id: "sdfaf",
  title: "コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ",
  author_id: "fsdaf",
  author_name: "ワイン大好き太郎",
  image_src: dummyImage.src,
  price: 2000,
  categories: categoryDummyData[0],
  tags: "樽熟成",
  sober_or_sweet: 3,
  tart: 3,
  fruity: 3,
  description:
    "チリのワインといえばコノスル！\nカベルネソーヴィニヨンの中でも特におすすめの一本です。",
  place: "スーパー,ネット通販",
  erudition:
    "カベルネソーヴィニヨンはボルドー地方の赤ワインの原料としても有名です。",
  category_id: 1,
  created_at: "2021-10-10",
};

export const articleDummyData: ArticleProps = {
  title: "コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ",
  author_name: "ワイン大好き太郎",
  tags: "樽熟成",
  sober_or_sweet: 3,
  tart: 3,
  fruity: 3,
  categories: categoryDummyData[0],
  id: "1",
  price: 1500,
};
