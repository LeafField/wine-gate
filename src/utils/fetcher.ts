import { supabase } from "./supabase";
import { EditingPageSchemaType, ImageInputSchemaType } from "./schema";
import { User } from "../store";
import { revalidateServer } from "../server/revalidate";

export const postWine = async ({
  value,
  user,
  image,
}: {
  value: EditingPageSchemaType;
  user: User;
  image: ImageInputSchemaType;
}) => {
  if (!image) throw new Error("画像が必須です");
  const IMAGE_SRC = `${user.id}/${image.name}`;

  const { error: imageError, data: imageData } = await supabase.storage
    .from("wines")
    .upload(IMAGE_SRC, image);

  if (imageError) throw new Error("画像のアップロードに失敗しました");
  if (imageData) {
    const { data, error } = await supabase
      .from("wines")
      .insert({
        title: value.title.replace(/　/g, " "),
        price: value.price,
        place: value.place,
        description: value.description,
        erudition: value.erudition,
        category_id: Number(value.category_id),
        fruity: value.fruity,
        tart: value.tart,
        sober_or_sweet: value.sober_or_sweet,
        tags: value.tags.join(),
        author_id: user.id,
        author_name: user.username,
        image_src: imageData.path,
      })
      .select();
    if (error) throw new Error("投稿に失敗しました");
    return data[0].id;
  }
};

export const updateWine = async ({
  id,
  image,
  oldWineImage,
  user_id,
  value,
}: {
  id: string;
  user_id: string;
  value: EditingPageSchemaType;
  image: ImageInputSchemaType | null;
  oldWineImage: string;
}) => {
  revalidateServer(id);
  if (image) {
    const IMAGE_SRC = `${user_id}/${image.name}`;
    const { error } = await supabase.storage
      .from("wines")
      .upload(IMAGE_SRC, image);
    if (error) throw new Error("画像のアップロードに失敗しました");
    const { error: deleteError } = await supabase.storage
      .from("wines")
      .remove([oldWineImage]);
    if (deleteError) throw new Error("画像の削除に失敗しました");
  }
  const { data, error } = await supabase
    .from("wines")
    .update({
      title: value.title.replace(/　/g, " "),
      price: value.price,
      place: value.place,
      description: value.description,
      erudition: value.erudition,
      category_id: Number(value.category_id),
      fruity: value.fruity,
      tart: value.tart,
      sober_or_sweet: value.sober_or_sweet,
      tags: value.tags.join(),
      image_src: image ? `${user_id}/${image.name}` : oldWineImage,
    })
    .eq("id", id)
    .select();
  if (error) throw new Error("ワインの更新に失敗しました");
  return data[0].id;
};

export const favoriteCount = async (wine_id: string) => {
  const { count, error } = await supabase
    .from("favorite")
    .select("id", { count: "exact", head: true })
    .eq("wine_id", wine_id);
  if (error) throw new Error("お気に入りのカウントに失敗しました");
  return count;
};

export const favoriteUserOne = async (wine_id: string) => {
  const { data, error } = await supabase
    .from("favorite")
    .select("user_id")
    .eq("wine_id", wine_id);
  if (error) throw new Error("お気に入りのユーザー取得に失敗しました");
  return data[0] ? data[0] : null;
};

export const addFavorite = async ({
  wine_id,
  user_id,
}: {
  wine_id: string;
  user_id: string;
}) => {
  const { error } = await supabase
    .from("favorite")
    .insert({ wine_id, user_id });

  if (error) throw new Error("お気に入りの登録に失敗しました");
};

export const removeFavorite = async ({
  wine_id,
  user_id,
}: {
  wine_id: string;
  user_id: string;
}) => {
  const { error } = await supabase
    .from("favorite")
    .delete()
    .eq("wine_id", wine_id)
    .eq("user_id", user_id);

  if (error) throw new Error("お気に入りの解除に失敗しました");
};

export const getWine = async (id: string) => {
  const { data, error } = await supabase
    .from("wines")
    .select("*,categories(*)")
    .eq("id", id)
    .single();
  if (error) throw new Error("ワインの取得に失敗しました");

  return data;
};
