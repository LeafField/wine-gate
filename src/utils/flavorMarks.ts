import { ReactNode } from "react";

type Marks = {
  value: number;
  label: ReactNode;
};

const sweet: Marks[] = [
  {
    value: 1,
    label: "辛口",
  },
  {
    value: 2,
    label: "中辛",
  },
  {
    value: 3,
    label: "中口",
  },
  {
    value: 4,
    label: "中甘",
  },
  {
    value: 5,
    label: "甘口",
  },
];

export const tart: Marks[] = [
  {
    value: 1,
    label: "ない",
  },
  {
    value: 2,
    label: "弱い",
  },
  {
    value: 3,
    label: "普通",
  },
  {
    value: 4,
    label: "強め",
  },
  {
    value: 5,
    label: "酸っぱい",
  },
];
