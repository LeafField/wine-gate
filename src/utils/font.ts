import { Noto_Sans_JP, Kumar_One } from "next/font/google";

export const notoStyle = Noto_Sans_JP({ subsets: ["latin"] });
export const KumarOne = Kumar_One({ subsets: ["latin"], weight: "400" });
