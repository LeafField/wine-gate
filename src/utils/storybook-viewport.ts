export const mobileViewport = {
  viewport: {
    defaultViewport: "iphone6",
  },
};
