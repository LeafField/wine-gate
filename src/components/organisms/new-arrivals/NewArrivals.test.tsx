import { render, screen } from "@testing-library/react";
import NewArrivals from "./NewArrivals";
import { smallArticleDummyData } from "../../../utils/dummyData";

describe("NewArrivalsの結合テスト", () => {
  test("コンポーネントがクラッシュせずにレンダリングされる", () => {
    render(<NewArrivals articles={smallArticleDummyData} />);
    const sectionElement = screen.getByRole("heading", { name: "新着ワイン" });
    expect(sectionElement).toBeInTheDocument();
  });

  test("正しい数のSmallArticleコンポーネントがレンダリングされる", () => {
    render(<NewArrivals articles={smallArticleDummyData} />);
    const smallArticleElements = screen.getAllByRole("article");
    expect(smallArticleElements.length).toEqual(3);
  });

  test("スナップショットテスト", () => {
    const { container } = render(
      <NewArrivals articles={smallArticleDummyData} />,
    );
    expect(container).toMatchSnapshot();
  });
});
