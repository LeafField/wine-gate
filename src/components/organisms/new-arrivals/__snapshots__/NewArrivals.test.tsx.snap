// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewArrivalsの結合テスト スナップショットテスト 1`] = `
<div>
  <section
    class="flex w-full flex-col gap-[3.75rem] px-6 tablet:w-[20.75rem] tablet:px-0 laptop:w-[43.0625rem] desktop:w-[64.9375rem]"
  >
    <div
      class="col-span-3"
    >
      <h2
        class="text-heading2 border-b-2 border-red  "
      >
        新着ワイン
      </h2>
    </div>
    <div
      class="grid grid-cols-1 justify-items-stretch gap-5 laptop:grid-cols-2 desktop:grid-cols-3"
    >
      <a
        class="group row-span-3 grid h-full w-full grid-rows-subgrid gap-y-0 border border-gray transition-colors duration-500 hover:border-blue"
        href="1"
        role="article"
      >
        <div
          class="aspect-video w-full overflow-hidden bg-gray "
        >
          <figure
            class="relative h-full w-full transition-transform duration-300 group-hover:scale-110 "
          >
            <img
              alt="ワイン画像"
              data-nimg="fill"
              decoding="async"
              fetchpriority="high"
              loading="eager"
              sizes="(min-width:768px) 332px,100vw"
              src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=16&q=75 16w, /_next/image?url=%2Fimg.jpg&w=32&q=75 32w, /_next/image?url=%2Fimg.jpg&w=48&q=75 48w, /_next/image?url=%2Fimg.jpg&w=64&q=75 64w, /_next/image?url=%2Fimg.jpg&w=96&q=75 96w, /_next/image?url=%2Fimg.jpg&w=128&q=75 128w, /_next/image?url=%2Fimg.jpg&w=256&q=75 256w, /_next/image?url=%2Fimg.jpg&w=384&q=75 384w, /_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: contain; color: transparent;"
            />
          </figure>
        </div>
        <h3
          class="w-full px-2 text-base text-black"
        >
          コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ
        </h3>
        <div
          class="mt-auto flex flex-row justify-center gap-2 pb-2 pt-2"
        >
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-sm text-black "
          >
            渋い
          </div>
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-sm text-black "
          >
            樽熟成
          </div>
        </div>
      </a>
      <a
        class="group row-span-3 grid h-full w-full grid-rows-subgrid gap-y-0 border border-gray transition-colors duration-500 hover:border-blue"
        href="2"
        role="article"
      >
        <div
          class="aspect-video w-full overflow-hidden bg-gray "
        >
          <figure
            class="relative h-full w-full transition-transform duration-300 group-hover:scale-110 "
          >
            <img
              alt="ワイン画像"
              data-nimg="fill"
              decoding="async"
              fetchpriority="high"
              loading="eager"
              sizes="(min-width:768px) 332px,100vw"
              src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=16&q=75 16w, /_next/image?url=%2Fimg.jpg&w=32&q=75 32w, /_next/image?url=%2Fimg.jpg&w=48&q=75 48w, /_next/image?url=%2Fimg.jpg&w=64&q=75 64w, /_next/image?url=%2Fimg.jpg&w=96&q=75 96w, /_next/image?url=%2Fimg.jpg&w=128&q=75 128w, /_next/image?url=%2Fimg.jpg&w=256&q=75 256w, /_next/image?url=%2Fimg.jpg&w=384&q=75 384w, /_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: contain; color: transparent;"
            />
          </figure>
        </div>
        <h3
          class="w-full px-2 text-base text-black"
        >
          コノスル シャルドネ ビシクレタ・レゼルバ
        </h3>
        <div
          class="mt-auto flex flex-row justify-center gap-2 pb-2 pt-2"
        >
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-sm text-black "
          >
            辛口白
          </div>
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-sm text-black "
          >
            フルーティ
          </div>
        </div>
      </a>
      <a
        class="group row-span-3 grid h-full w-full grid-rows-subgrid gap-y-0 border border-gray transition-colors duration-500 hover:border-blue"
        href="3"
        role="article"
      >
        <div
          class="aspect-video w-full overflow-hidden bg-gray "
        >
          <figure
            class="relative h-full w-full transition-transform duration-300 group-hover:scale-110 "
          >
            <img
              alt="ワイン画像"
              data-nimg="fill"
              decoding="async"
              fetchpriority="high"
              loading="eager"
              sizes="(min-width:768px) 332px,100vw"
              src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=16&q=75 16w, /_next/image?url=%2Fimg.jpg&w=32&q=75 32w, /_next/image?url=%2Fimg.jpg&w=48&q=75 48w, /_next/image?url=%2Fimg.jpg&w=64&q=75 64w, /_next/image?url=%2Fimg.jpg&w=96&q=75 96w, /_next/image?url=%2Fimg.jpg&w=128&q=75 128w, /_next/image?url=%2Fimg.jpg&w=256&q=75 256w, /_next/image?url=%2Fimg.jpg&w=384&q=75 384w, /_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: contain; color: transparent;"
            />
          </figure>
        </div>
        <h3
          class="w-full px-2 text-base text-black"
        >
          コノスル カルメネール ビシクレタ・レゼルバ
        </h3>
        <div
          class="mt-auto flex flex-row justify-center gap-2 pb-2 pt-2"
        >
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-sm text-black "
          >
            酸っぱい
          </div>
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-sm text-black "
          >
            樽熟成
          </div>
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-sm text-black "
          >
            濃厚
          </div>
        </div>
      </a>
    </div>
  </section>
</div>
`;
